cpp/memilio/io/mobility_io.h

```diff
@@ -60,6 +62,129 @@ IOResult<Eigen::MatrixXd> read_mobility_formatted(const std::string& filename);
+        js_node["Model"]   = js_node_model;
+        auto node_filename = path_join(abs_path, "GraphNode" + std::to_string(inode) + ".json");
+        BOOST_OUTCOME_TRY(write_json(node_filename, js_node));
+
+        //list of edges
+        auto out_edges = graph.out_edges(inode);
+        if (out_edges.size()) {
+            Json::Value js_edges(Json::arrayValue);
+            for (auto& e : graph.out_edges(inode)) {
+                BOOST_OUTCOME_TRY(js_edge_params, serialize_json(e.property, ioflags));
+                Json::Value js_edge{Json::objectValue};
+                js_edge["StartNodeIndex"] = Json::UInt64(e.start_node_idx);
+                js_edge["EndNodeIndex"]   = Json::UInt64(e.end_node_idx);
+                js_edge["Parameters"]     = js_edge_params;
+                js_edges.append(std::move(js_edge));
+            }
+            auto edge_filename = path_join(abs_path, "GraphEdges_node" + std::to_string(inode) + ".json");
+            BOOST_OUTCOME_TRY(write_json(edge_filename, js_edges));
+        }
+    }
+
+    return success();
+}
+
+/**
+ * @brief reads graph json files and returns a simulation graph.
+ * See write_graph for information of expected files.
+ * @tparam the type of the simulation model.
+ * @param directory directory from where graph should be read.
+ */
+template <class Model>
+IOResult<Graph<Model, MigrationParameters>> read_graph(const std::string& directory, int ioflags = IOF_None)
+{
+    std::string abs_path;
+    if (!file_exists(directory, abs_path)) {
+        log_error("Directory {} does not exist.", directory);
+        return failure(StatusCode::FileNotFound, directory);
+    }
+
+    auto graph = Graph<Model, MigrationParameters>{};
+
+    //read nodes, as many as files are available
+    for (auto inode = 0;; ++inode) {
+        auto node_filename = path_join(abs_path, "GraphNode" + std::to_string(inode) + ".json");
+        if (!file_exists(node_filename, node_filename)) {
+            break;
+        }
+        BOOST_OUTCOME_TRY(js_node, read_json(node_filename));
+        if (!js_node["NodeId"].isInt()) {
+            log_error("NodeId field must be an integer.");
+            return failure(StatusCode::InvalidType, node_filename + ", NodeId must be an integer.");
+        }
+        auto node_id = js_node["NodeId"].asInt();
+        BOOST_OUTCOME_TRY(model, deserialize_json(js_node["Model"], Tag<Model>{}, ioflags));
+        graph.add_node(node_id, model);
+    }
+
+    //read edges; nodes must already be available for that)
+    for (auto inode = size_t(0); inode < graph.nodes().size(); ++inode) {
+        //list of edges
+        auto edge_filename = path_join(abs_path, "GraphEdges_node" + std::to_string(inode) + ".json");
+        BOOST_OUTCOME_TRY(js_edges, read_json(edge_filename));
+
+        for (auto& e : js_edges) {
+            auto start_node_idx  = inode;
+            auto js_end_node_idx = e["EndNodeIndex"];
+            if (!js_end_node_idx.isUInt64()) {
```

Comment:
the json file format doesn't differentiate between 32 or 64 bit integers, only C++ has to differentiate. I guess jsoncpp does a bounds check somewhere to see if the value in the file fits in the requested type. If it was a negative value, it probably would throw an exception or fail an assertion. But negative indices would be invalid, so that's ok I think.